docs/experiments/join-mode.md

```diff
@@ -0,0 +1,197 @@
+---
+slug: join-mode
+append_help_link: true
+description: "Join mode runs several Semgrep rules at once and only returns results if certain conditions on the results are met."
+---
+
+# Join Mode
+
+Join mode runs several Semgrep rules at once and only returns results if certain conditions on the results are met. Semgrep is brilliant for finding code patterns with an easy syntax, but its search is typically limited to single files. Join mode is an experimental mode that lets you cross file boundaries, allowing you to write rules for whole code bases instead of individual files. As the name implies, this was inspired by join clauses in SQL queries.
+
+You can think of join mode like this: distinct Semgrep rules are used to gather information about a code base. Then, conditions you define are used to select specific results from these rules, and the selected results are reported by Semgrep in the normal fashion. You can join results on metavariable contents or on the result's file path.
+
+## Example
+
+Here is an example join mode rule that detects a cross-site scripting (XSS) vulnerability with high precision.
+
+```yaml
+rules:
+- id: flask-likely-xss
+  mode: join
+  join:
+    refs:
+      - rule: flask-user-input.yaml
+        as: user-input
+      - rule: unescaped-template-extension.yaml
+        as: unescaped-extensions
+      - rule: any-template-var.yaml
+        renames:
+        - from: '$...EXPR'
+          to: '$VAR'
+        as: template-vars
+    on:
+    - 'user-input.$VAR == unescaped-extensions.$VALUE'
+    - 'unescaped-extensions.$VAR == template-vars.$VAR'
+    - 'unescaped-extensions.$PATH > template-vars.path'
+  message: |
+    Detected a XSS vulnerability: '$VAR' is rendered
+    unsafely in '$PATH'. 
+  severity: ERROR
+```
+
+Let's explore how this works. First, some background on the vulnerability. Second, we'll walk through the join mode rule.
+
+**Vulnerability Background**
```

Comment:
```suggestion
**Vulnerability background**
```

Also, any particular reason to use bold here (and below) instead of a smaller heading? The docs’ search indexing works in part on grabbing hierarchy from the headers.

Reply:
I didn't think it merited its own section because it didn't quite match the "tree structure" I had in my head. If you think it's more beneficial to break this out, though, I can do that!

Reply:
The search will index all the text regardless, it just might rank things differently because of the headings. Since this is v1 of the docs for this feature, it’s probably a-ok as-is.